homeassistant/components/intellifire/config_flow.py

```diff
@@ -71,47 +78,117 @@ async def _find_fireplaces(self):
         LOGGER.debug("Configured Hosts: %s", configured_hosts)
         LOGGER.debug("Not Configured Hosts: %s", self._not_configured_hosts)
 
-    async def _async_validate_and_create_entry(self, host: str) -> FlowResult:
-        """Validate and create the entry."""
+    async def validate_api_access_and_create_or_update(
+        self, *, host: str, username: str, password: str, serial: str
+    ):
+        """Validate username/password against api."""
+        ift_control = IntellifireControlAsync(fireplace_ip=host)
+
+        LOGGER.debug("Attempting login to iftapi with: %s", username)
+        # This can throw an error which will be handled above
+        try:
+            await ift_control.login(username=username, password=password)
+            await ift_control.get_username()
+        finally:
+            await ift_control.close()
+
+        data = {CONF_HOST: host, CONF_PASSWORD: password, CONF_USERNAME: username}
+
+        # Update or Create
+        existing_entry = await self.async_set_unique_id(serial)
+        if existing_entry:
+            self.hass.config_entries.async_update_entry(existing_entry, data=data)
+            await self.hass.config_entries.async_reload(existing_entry.entry_id)
+            return self.async_abort(reason="reauth_successful")
+        return self.async_create_entry(title=f"Fireplace {serial}", data=data)
+
+    async def async_step_api_config(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Configure API access."""
+
+        errors = {}
+        control_schema = vol.Schema(
+            {
+                vol.Required(CONF_USERNAME): str,
+                vol.Required(CONF_PASSWORD): str,
+            }
+        )
+
+        if user_input is not None:
+
+            control_schema = vol.Schema(
+                {
+                    vol.Required(
+                        CONF_USERNAME, default=user_input.get(CONF_USERNAME, "")
+                    ): str,
+                    vol.Required(
+                        CONF_PASSWORD, default=user_input.get(CONF_PASSWORD, "")
+                    ): str,
+                }
+            )
+
+            if user_input[CONF_USERNAME] != "":
```

Comment:
This check is a bit weird. The frontend won't allow an empty string for required form items. Also if an empty string is invalid we'd want to show an error in the form when showing it again to the user. Otherwise there's no feedback to the user what went wrong. But as said, normally we don't implement this since the frontend already guards for it.